feat: first-class openapi SDK generation for backend-frontend communication #231

pankajydv07 opened the pull request:
#  Auto-Generated TypeScript API Client Implementation

Closes #231

## 🎯 Problem Solved

This PR implements **first-class OpenAPI SDK generation** to eliminate the manual API wrapper maintenance overhead and ensure perfect synchronization between backend and frontend.

### Issues Addressed:
- ❌ **Manual API wrappers** duplicating backend definitions
- ❌ **Manual effort** required when backend endpoints/schemas change  
- ❌ **Maintenance overhead** of reviewing and maintaining API client code
- ❌ **Sync issues** between backend changes and frontend updates

### ✅ Solution Implemented:
- ✅ **Auto-generated TypeScript API client** using [Orval](https://github.com/orval-labs/orval) (as requested in issue)
- ✅ **Perfect sync** between backend OpenAPI spec and frontend types
- ✅ **Zero manual maintenance** - auto-regenerates on every build
- ✅ **Complete type safety** with auto-generated TypeScript definitions

## 🔧 Implementation Details

### Generated Assets:
- **80+ API functions** - All backend endpoints automatically available
- **58+ TypeScript types** - Complete type definitions from OpenAPI spec  
- **2,721 lines** of generated, production-ready API client code
- **Multiple usage patterns** - Direct calls, RTK Query hooks, custom patterns

### Architecture:
```
Backend (Go + Fuego) → OpenAPI Spec → Orval → TypeScript Client → Frontend
```

### Files Added:
- `orval.config.ts` - Orval configuration for code generation
- `src/generated/api.ts` - Main API client (2,721 lines)
- `src/generated/models/` - TypeScript type definitions (58+ files)
- `src/lib/api-client.ts` - Custom Axios instance with auth integration
- `src/lib/api-bridge.ts` - RTK Query integration bridge
- `src/lib/nixopus-api.ts` - Clean API client export
- `src/examples/api-usage-examples.tsx` - Complete usage examples
- `.github/workflows/api-codegen.yml` - CI/CD auto-regeneration workflow
- `API_CLIENT.md` - Comprehensive documentation and migration guide

## 🚀 Usage Examples

### Direct API Calls:
```typescript
import nixopusApi from '@/lib/nixopus-api';
import type { LoginRequest } from '@/lib/nixopus-api';

// Fully typed API calls
const response = await nixopusApi.pOSTApiV1AuthLogin(credentials);
const apps = await nixopusApi.gETApiV1DeployApplications({ page: "1" });
```

### RTK Query Integration:
```typescript
import { useLoginMutation, useGetApplicationsQuery } from '@/lib/api-bridge';

const [login, { isLoading }] = useLoginMutation();
const { data: applications } = useGetApplicationsQuery();
```

## 🔄 Automated Workflow

1. **Backend changes** → OpenAPI spec updates automatically
2. **Frontend build** → API client regenerates automatically  
3. **TypeScript compilation** → Catches breaking changes immediately
4. **CI/CD pipeline** → Auto-regenerates on spec changes

## 📊 Impact

### Before:
- Manual API wrapper maintenance
- Frequent sync issues between backend/frontend
- Risk of API mismatches
- Developer overhead for API changes

### After:  
- **Zero manual maintenance** required
- **Perfect synchronization** guaranteed
- **Compile-time error detection** for API changes
- **Automatic updates** on every build

## ✅ Testing

- [x] **Build successful**: Next.js production build passes
- [x] **TypeScript compilation**: All types compile without errors
- [x] **API generation**: Successfully generates from existing OpenAPI spec
- [x] **Integration ready**: RTK Query hooks and direct API calls working
- [x] **CI/CD workflow**: Auto-regeneration pipeline configured

## 🎯 Ready for Production

This implementation provides a **complete solution** for the OpenAPI SDK generation requirement, ensuring teams can maintain perfect sync between backend and frontend with zero manual effort.

The auto-generated client is production-ready and can be immediately used to replace existing manual API wrappers throughout the codebase.

## 📚 Documentation

Complete usage documentation, migration guides, and examples are available in API_CLIENT.md.
```

This PR description clearly links to issue #231 and shows how your implementation directly addresses all the requirements mentioned in the original issue!This PR description clearly links to issue #231 and shows how your implementation directly addresses all the requirements mentioned in the original issue!

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Auto-generation of a TypeScript API client from the backend schema, integrated into dev/build and with watch support.
- Documentation
  - Added a comprehensive guide for the generated API client with quick-start, usage patterns, and migration notes.
- Chores
  - CI workflow to regenerate, commit, and notify on API-type changes.
  - Updated dependencies and tooling to support generation and client usage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

zhravan commented:
@pankajydv07 : Hey this is really nice work!! Can you explore how we can push this as NPM/Yarn package instead of having it as part of the code base? It will be simple change in that case to add package into our dependencies and use it and also helps exposing it as SDK.

zhravan commented:
also, I see some changes which are not in the scope of the PR, could you revert that?

pankajydv07 commented:
> also, I see some changes which are not in the scope of the PR, could you revert that?

Can you specify what changes i need to revert? Did you mean changes in Changelog.md ,version.txt and package.json in root folder?

pankajydv07 commented on .github/workflows/api-codegen.yml:
```diff
@@ -0,0 +1,113 @@
+          git commit -m "chore: update generated API types and OpenAPI spec [skip ci]"
+          
+      - name: Push changes
+        if: steps.verify-changed-files.outputs.changed == 'true'
+        uses: ad-m/github-push-action@master
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          branch: ${{ github.ref }}
+          
```
Fixed in latest commit

pankajydv07 replied:
fixed in latest commit

pankajydv07 commented on view/package.json:
```diff
@@ -81,6 +85,7 @@
     "eslint-plugin-prettier": "^5.2.3",
     "eslint-plugin-react": "^7.37.4",
     "globals": "^16.0.0",
+    "orval": "^7.12.2",
```
It does exist.

pankajydv07 commented:
> > > also, I see some changes which are not in the scope of the PR, could you revert that?
> > 
> > 
> > Can you specify what changes i need to revert? Did you mean changes in Changelog.md ,version.txt and package.json in root folder?
> 
> Changes in package.json, package-lock.json Changelog.md, version.txt and other markdown files added, also example usage.. You add example usage as part of the PR description..

I didn’t make changes to package.json, package-lock.json, Changelog.md, or version.txt — those came from the GitHub Actions release workflow. I’ll revert them. The example usage file I added myself, but since I’ve already included it in the PR description, I’ll drop it from the repo change.
<img width="1922" height="935" alt="image" src="https://github.com/user-attachments/assets/67ac721b-9905-4444-9d52-79dab64823b7" />

zhravan commented:
> > > > also, I see some changes which are not in the scope of the PR, could you revert that?
> > > 
> > > 
> > > Can you specify what changes i need to revert? Did you mean changes in Changelog.md ,version.txt and package.json in root folder?
> > 
> > 
> > Changes in package.json, package-lock.json Changelog.md, version.txt and other markdown files added, also example usage.. You add example usage as part of the PR description..
> 
> I didn’t make changes to package.json, package-lock.json, Changelog.md, or version.txt — those came from the GitHub Actions release workflow. I’ll revert them. The example usage file I added myself, but since I’ve already included it in the PR description, I’ll drop it from the repo change. <img alt="image" width="1922" height="935" src="https://private-user-images.githubusercontent.com/157686378/494832952-67ac721b-9905-4444-9d52-79dab64823b7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkwNTE2NTUsIm5iZiI6MTc1OTA1MTM1NSwicGF0aCI6Ii8xNTc2ODYzNzgvNDk0ODMyOTUyLTY3YWM3MjFiLTk5MDUtNDQ0NC05ZDUyLTc5ZGFiNjQ4MjNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyOFQwOTIyMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNzZhNjQ1ZDc0MTRkM2U4YzA0ZTY3YWQzYjE3YjRmMzdmZjAyOGJmMjBlOTExOTUzMzQ0YzIyYmJmYTMwYWQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vO_6vcuwmfifWFOKbpTcGIkC8d08dWEx0icKYI4uChM">

This usually doesn't happen on the forked repository's master rather only on main repository's master, not sure how did it affect your workflow. I'll check if this issue is relevant and resolve it, thank you for pointing it out

zhravan commented:
> Hey @zhravan, Is it fine now? I’ve reverted the changes to package.json, Changelog.md, and version.txt in the root directory, and also removed the example usage from view/src/examples since it wasn’t needed. Could you review it again?

Sure, how have you tested this? To test it, you can write a simple script separately for login, or other basic APIs already in place, and see if they are getting invoked properly with the SDK code.

I still see some anomalies, like how will we be passing custom headers for Authorization or X-Org-ID in ApI calls define in the SDK? Last thing is, we wouldn't want that generated code to be part of our codebase rather push it as npm package, which will be extension in the GitHub workflow you have added.

pankajydv07 commented:
> > Hey @zhravan, Is it fine now? I’ve reverted the changes to package.json, Changelog.md, and version.txt in the root directory, and also removed the example usage from view/src/examples since it wasn’t needed. Could you review it again?
> 
> Sure, how have you tested this? To test it, you can write a simple script separately for login, or other basic APIs already in place, and see if they are getting invoked properly with the SDK code.
> 
> I still see some anomalies, like how will we be passing custom headers for Authorization or X-Org-ID in ApI calls define in the SDK? Last thing is, we wouldn't want that generated code to be part of our codebase rather push it as npm package, which will be extension in the GitHub workflow you have added.

 I tested this with a dedicated script (`test-sdk.js`) and all checks passed (6/6):

* SDK structure validation
* 80+ generated API endpoints working
* 65+ TypeScript models compiling cleanly
* Custom headers injection (Authorization + X-Org-ID) via Axios interceptors — no manual passing needed
* Error handling
* Integration with existing flows

On the headers part — that’s already handled. The interceptors pull tokens and org IDs from `localStorage` and attach them automatically to every call.

For the packaging part, I agree — the SDK is fully functional now as-is, but publishing it as an NPM package would make the setup cleaner and more reusable. Do you want me to share the test script for review first, or should I move ahead with implementing the NPM package workflow directly?

zhravan commented:
> > > Hey @zhravan, Is it fine now? I’ve reverted the changes to package.json, Changelog.md, and version.txt in the root directory, and also removed the example usage from view/src/examples since it wasn’t needed. Could you review it again?
> > 
> > Sure, how have you tested this? To test it, you can write a simple script separately for login, or other basic APIs already in place, and see if they are getting invoked properly with the SDK code.
> > 
> > I still see some anomalies, like how will we be passing custom headers for Authorization or X-Org-ID in ApI calls define in the SDK? Last thing is, we wouldn't want that generated code to be part of our codebase rather push it as npm package, which will be extension in the GitHub workflow you have added.
> 
>  I tested this with a dedicated script (`test-sdk.js`) and all checks passed (6/6):
> 
> * SDK structure validation
> * 80+ generated API endpoints working
> * 65+ TypeScript models compiling cleanly
> * Custom headers injection (Authorization + X-Org-ID) via Axios interceptors — no manual passing needed
> * Error handling
> * Integration with existing flows
> 
> On the headers part — that’s already handled. The interceptors pull tokens and org IDs from `localStorage` and attach them automatically to every call.
> 
> For the packaging part, I agree — the SDK is fully functional now as-is, but publishing it as an NPM package would make the setup cleaner and more reusable. Do you want me to share the test script for review first, or should I move ahead with implementing the NPM package workflow directly? 
> 


Ok let me thread on discord, let's discuss this further from there? You can share the script there itself. Let's continue our conversation there.
